File: grpc/nodejs/test/unit/node.test.js
/**
 * Unit tests for Node class.
 * All tests are mocked — no Freeplane server required.
 */

const Node = require('../../src/node');
const { NodeNotFoundError } = require('../../src/exceptions');

describe('Node', () => {
  let mockClient;
  let mockMindmap;

  beforeEach(() => {
    mockClient = {
      getNodeText: jest.fn(),
      setNodeText: jest.fn(),
      createChild: jest.fn(),
      listChildNodes: jest.fn(),
      getParentNode: jest.fn(),
      deleteChild: jest.fn(),
      moveNode: jest.fn(),
      groovy: jest.fn(),
      nodeColorSet: jest.fn(),
      nodeBackgroundColorSet: jest.fn(),
      getNodeNote: jest.fn(),
      nodeNoteSet: jest.fn(),
      nodeAttributeAdd: jest.fn(),
      getNodeLink: jest.fn(),
      nodeLinkSet: jest.fn(),
      nodeTagSet: jest.fn(),
      nodeTagAdd: jest.fn(),
      nodeAddIcon: jest.fn(),
      focusNode: jest.fn(),
    };
    mockMindmap = { _nodeId: 'root1' };
  });

  describe('constructor and properties', () => {
    test('stores client, nodeId, and mindmap', () => {
      const node = new Node(mockClient, 'node1', mockMindmap);
      expect(node.client).toBe(mockClient);
      expect(node.nodeId).toBe('node1');
      expect(node.mindmap).toBe(mockMindmap);
    });

    test('mindmap can be undefined', () => {
      const node = new Node(mockClient, 'node1');
      expect(node.mindmap).toBeUndefined();
    });
  });

  describe('getText', () => {
    test('returns text from client response', async () => {
      mockClient.getNodeText.mockResolvedValue({ text: 'Hello World' });
      const node = new Node(mockClient, 'node1');
      const text = await node.getText();
      expect(text).toBe('Hello World');
      expect(mockClient.getNodeText).toHaveBeenCalledWith('node1');
    });
  });

  describe('setText', () => {
    test('calls client.setNodeText with nodeId and text', async () => {
      mockClient.setNodeText.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      await node.setText('New Text');
      expect(mockClient.setNodeText).toHaveBeenCalledWith('node1', 'New Text');
    });
  });

  describe('addChild', () => {
    test('returns a new Node with the created child id', async () => {
      mockClient.createChild.mockResolvedValue({ nodeId: 'child1', nodeText: 'Child' });
      const node = new Node(mockClient, 'node1', mockMindmap);
      const child = await node.addChild('Child');
      expect(child.nodeId).toBe('child1');
      expect(child.client).toBe(mockClient);
      expect(child.mindmap).toBe(mockMindmap);
      expect(mockClient.createChild).toHaveBeenCalledWith('Child', 'node1');
    });

    test('updates mindmap nodeId when mindmap exists', async () => {
      mockClient.createChild.mockResolvedValue({ nodeId: 'child1', nodeText: 'Child' });
      const node = new Node(mockClient, 'node1', mockMindmap);
      await node.addChild('Child');
      expect(mockMindmap._nodeId).toBe('child1');
    });
  });

  describe('children', () => {
    test('returns array of Node instances', async () => {
      mockClient.listChildNodes.mockResolvedValue({
        success: true,
        children: [
          { nodeId: 'c1', text: 'Child 1' },
          { nodeId: 'c2', text: 'Child 2' },
        ],
      });
      const node = new Node(mockClient, 'node1', mockMindmap);
      const children = await node.children();
      expect(children).toHaveLength(2);
      expect(children[0].nodeId).toBe('c1');
      expect(children[1].nodeId).toBe('c2');
    });

    test('returns empty array when no children', async () => {
      mockClient.listChildNodes.mockResolvedValue({ success: true, children: [] });
      const node = new Node(mockClient, 'node1');
      const children = await node.children();
      expect(children).toHaveLength(0);
    });
  });

  describe('parent', () => {
    test('returns parent Node', async () => {
      mockClient.getParentNode.mockResolvedValue({
        success: true,
        parentNodeId: 'parent1',
        parentNodeText: 'Parent',
      });
      const node = new Node(mockClient, 'node1', mockMindmap);
      const parent = await node.parent();
      expect(parent.nodeId).toBe('parent1');
    });

    test('throws NodeNotFoundError when no parent', async () => {
      mockClient.getParentNode.mockResolvedValue({
        success: true,
        parentNodeId: '',
      });
      const node = new Node(mockClient, 'node1');
      await expect(node.parent()).rejects.toThrow(NodeNotFoundError);
    });
  });

  describe('delete', () => {
    test('returns success from client', async () => {
      mockClient.deleteChild.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.delete();
      expect(result).toBe(true);
      expect(mockClient.deleteChild).toHaveBeenCalledWith('node1');
    });
  });

  describe('move', () => {
    test('returns success from client', async () => {
      mockClient.moveNode.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.move('newParent');
      expect(result).toBe(true);
      expect(mockClient.moveNode).toHaveBeenCalledWith('node1', 'newParent');
    });
  });

  describe('setStyle', () => {
    test('calls groovy and returns true on success', async () => {
      mockClient.groovy.mockResolvedValue('');
      const node = new Node(mockClient, 'node1');
      const result = await node.setStyle('bubble');
      expect(result).toBe(true);
      expect(mockClient.groovy).toHaveBeenCalled();
    });

    test('returns false when groovy returns error', async () => {
      mockClient.groovy.mockResolvedValue('Error: something went wrong');
      const node = new Node(mockClient, 'node1');
      const result = await node.setStyle('bubble');
      expect(result).toBe(false);
    });
  });

  describe('setColor', () => {
    test('calls client.nodeColorSet', async () => {
      mockClient.nodeColorSet.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.setColor(255, 0, 0, 255);
      expect(result).toBe(true);
      expect(mockClient.nodeColorSet).toHaveBeenCalledWith('node1', 255, 0, 0, 255);
    });
  });

  describe('setBackgroundColor', () => {
    test('calls client.nodeBackgroundColorSet', async () => {
      mockClient.nodeBackgroundColorSet.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.setBackgroundColor(0, 255, 0);
      expect(result).toBe(true);
    });
  });

  describe('getNote', () => {
    test('returns note from client response', async () => {
      mockClient.getNodeNote.mockResolvedValue({ note: 'My note', hasNote: true });
      const node = new Node(mockClient, 'node1');
      const note = await node.getNote();
      expect(note).toBe('My note');
    });
  });

  describe('setNote', () => {
    test('returns success from client', async () => {
      mockClient.nodeNoteSet.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.setNote('My note');
      expect(result).toBe(true);
    });
  });

  describe('setAttribute', () => {
    test('returns success from client', async () => {
      mockClient.nodeAttributeAdd.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.setAttribute('key', 'value');
      expect(result).toBe(true);
    });
  });

  describe('setAttributes', () => {
    test('sets multiple attributes', async () => {
      mockClient.nodeAttributeAdd.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.setAttributes({ key1: 'val1', key2: 'val2' });
      expect(result).toBe(true);
      expect(mockClient.nodeAttributeAdd).toHaveBeenCalledTimes(2);
    });
  });

  describe('getLinks', () => {
    test('returns array with link when hasLink is true', async () => {
      mockClient.getNodeLink.mockResolvedValue({ link: 'http://example.com', hasLink: true });
      const node = new Node(mockClient, 'node1');
      const links = await node.getLinks();
      expect(links).toEqual(['http://example.com']);
    });

    test('returns empty array when hasLink is false', async () => {
      mockClient.getNodeLink.mockResolvedValue({ link: '', hasLink: false });
      const node = new Node(mockClient, 'node1');
      const links = await node.getLinks();
      expect(links).toEqual([]);
    });
  });

  describe('setTags', () => {
    test('returns success from client', async () => {
      mockClient.nodeTagSet.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.setTags(['tag1', 'tag2']);
      expect(result).toBe(true);
    });
  });

  describe('addTags', () => {
    test('returns success from client', async () => {
      mockClient.nodeTagAdd.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.addTags(['tag1']);
      expect(result).toBe(true);
    });
  });

  describe('addIcon', () => {
    test('returns success from client', async () => {
      mockClient.nodeAddIcon.mockResolvedValue({ success: true });
      const node = new Node(mockClient, 'node1');
      const result = await node.addIcon('flag');
      expect(result).toBe(true);
    });
  });

  describe('select', () => {
    test('delegates to client.focusNode', async () => {
      mockClient.focusNode.mockResolvedValue(true);
      const node = new Node(mockClient, 'node1');
      const result = await node.select();
      expect(result).toBe(true);
      expect(mockClient.focusNode).toHaveBeenCalledWith('node1');
    });
  });

  describe('center', () => {
    test('delegates to client.focusNode', async () => {
      mockClient.focusNode.mockResolvedValue(true);
      const node = new Node(mockClient, 'node1');
      const result = await node.center();
      expect(result).toBe(true);
    });
  });

  describe('refresh', () => {
    test('calls getText to reload state', async () => {
      mockClient.getNodeText.mockResolvedValue({ text: 'updated' });
      const node = new Node(mockClient, 'node1');
      await node.refresh();
      expect(mockClient.getNodeText).toHaveBeenCalledWith('node1');
    });
  });

  describe('toString', () => {
    test('returns formatted string', () => {
      const node = new Node(mockClient, 'node1');
      expect(node.toString()).toBe('Node(id=node1)');
    });
  });
});

File: grpc/golang/tests/node_test.go
package tests

import (
	"context"
	"testing"

	freeplane "github.com/metacoma/freeplane_plugin_grpc/grpc/golang"
	"go.uber.org/mock/gomock"
	mockpb "github.com/metacoma/freeplane_plugin_grpc/grpc/golang/internal/mock"
	pb "github.com/metacoma/freeplane_plugin_grpc/grpc/golang/pb"
)

func TestNode_Client(t *testing.T) {
	client := freeplane.NewFreeplaneClient("127.0.0.1", 50051)
	node := freeplane.NewNode(client, "node-1", nil)
	if node.Client() != client {
		t.Error("Client() returned wrong client")
	}
}

func TestNode_NodeID(t *testing.T) {
	client := freeplane.NewFreeplaneClient("127.0.0.1", 50051)
	node := freeplane.NewNode(client, "node-123", nil)
	if node.NodeID() != "node-123" {
		t.Errorf("NodeID() = %s, want node-123", node.NodeID())
	}
}

func TestNode_MindMap(t *testing.T) {
	client := freeplane.NewFreeplaneClient("127.0.0.1", 50051)
	mm := freeplane.NewMindMap(client, "map-1", "node-1")
	node := freeplane.NewNode(client, "node-1", mm)
	if node.MindMap() != mm {
		t.Error("MindMap() returned wrong mindmap")
	}
}

func TestNode_String(t *testing.T) {
	// Node.String() calls GetText which requires a connection.
	// We test that the method exists and has the correct signature.
	var node *freeplane.Node
	_ = node.String // suppress unused warning
}

func TestMockNode_AddChild(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().CreateChild(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.CreateChildResponse{NodeId: "child-1", NodeText: "Child"}, nil)

	// We can't directly inject the mock, but we can verify the interface works
	_, err := mockClient.CreateChild(context.Background(), &pb.CreateChildRequest{
		Name:         "Child",
		ParentNodeId: "parent-1",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_GetText(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().GetNodeText(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.GetNodeTextResponse{Success: true, Text: "Hello World"}, nil)

	_, err := mockClient.GetNodeText(context.Background(), &pb.GetNodeTextRequest{NodeId: "node-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_SetText(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().SetNodeText(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.SetNodeTextResponse{Success: true}, nil)

	_, err := mockClient.SetNodeText(context.Background(), &pb.SetNodeTextRequest{NodeId: "node-1", Text: "New Text"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_Children(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().ListChildNodes(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.ListChildNodesResponse{
		Success: true,
		Children: []*pb.ChildNodeInfo{
			{NodeId: "c1", Text: "Child 1"},
			{NodeId: "c2", Text: "Child 2"},
		},
	}, nil)

	resp, err := mockClient.ListChildNodes(context.Background(), &pb.ListChildNodesRequest{NodeId: "parent-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if len(resp.GetChildren()) != 2 {
		t.Errorf("expected 2 children, got %d", len(resp.GetChildren()))
	}
}

func TestMockNode_Parent(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().GetParentNode(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.GetParentNodeResponse{
		Success:        true,
		ParentNodeId:   "parent-1",
		ParentNodeText: "Parent",
	}, nil)

	resp, err := mockClient.GetParentNode(context.Background(), &pb.GetParentNodeRequest{NodeId: "child-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if resp.GetParentNodeId() != "parent-1" {
		t.Errorf("expected parent_node_id parent-1, got %s", resp.GetParentNodeId())
	}
}

func TestMockNode_Parent_NoParent(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().GetParentNode(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.GetParentNodeResponse{
		Success:        false,
		ParentNodeId:   "",
		ParentNodeText: "",
	}, nil)

	resp, err := mockClient.GetParentNode(context.Background(), &pb.GetParentNodeRequest{NodeId: "root"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if resp.GetSuccess() {
		t.Error("expected success=false for root node")
	}
}

func TestMockNode_Delete(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().DeleteChild(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.DeleteChildResponse{Success: true}, nil)

	_, err := mockClient.DeleteChild(context.Background(), &pb.DeleteChildRequest{NodeId: "node-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_Move(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().MoveNode(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.MoveNodeResponse{Success: true}, nil)

	_, err := mockClient.MoveNode(context.Background(), &pb.MoveNodeRequest{NodeId: "node-1", NewParentNodeId: "node-2"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_SetNote(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeNoteSet(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeNoteSetResponse{Success: true}, nil)

	_, err := mockClient.NodeNoteSet(context.Background(), &pb.NodeNoteSetRequest{NodeId: "node-1", Note: "A note"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_GetNote(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().GetNodeNote(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.GetNodeNoteResponse{Success: true, HasNote: true, Note: "A note"}, nil)

	_, err := mockClient.GetNodeNote(context.Background(), &pb.GetNodeNoteRequest{NodeId: "node-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_SetAttribute(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeAttributeAdd(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeAttributeAddResponse{Success: true}, nil)

	_, err := mockClient.NodeAttributeAdd(context.Background(), &pb.NodeAttributeAddRequest{
		NodeId: "node-1", AttributeName: "key", AttributeValue: "value",
	})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_SetLinks(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeLinkSet(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeLinkSetResponse{Success: true}, nil)

	_, err := mockClient.NodeLinkSet(context.Background(), &pb.NodeLinkSetRequest{NodeId: "node-1", Link: "http://example.com"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_GetLinks(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().GetNodeLink(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.GetNodeLinkResponse{Success: true, HasLink: true, Link: "http://example.com"}, nil)

	_, err := mockClient.GetNodeLink(context.Background(), &pb.GetNodeLinkRequest{NodeId: "node-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_SetTags(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeTagSet(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeTagSetResponse{Success: true}, nil)

	_, err := mockClient.NodeTagSet(context.Background(), &pb.NodeTagSetRequest{NodeId: "node-1", Tags: []string{"tag1", "tag2"}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_AddTags(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeTagAdd(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeTagAddResponse{Success: true}, nil)

	_, err := mockClient.NodeTagAdd(context.Background(), &pb.NodeTagAddRequest{NodeId: "node-1", Tags: []string{"tag1"}})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_AddIcon(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeAddIcon(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeAddIconResponse{Success: true}, nil)

	_, err := mockClient.NodeAddIcon(context.Background(), &pb.NodeAddIconRequest{NodeId: "node-1", IconName: "star"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_SetColor(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeColorSet(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeColorSetResponse{Success: true}, nil)

	_, err := mockClient.NodeColorSet(context.Background(), &pb.NodeColorSetRequest{NodeId: "node-1", Red: 255, Green: 0, Blue: 0, Alpha: 255})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_SetBackgroundColor(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().NodeBackgroundColorSet(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.NodeBackgroundColorSetResponse{Success: true}, nil)

	_, err := mockClient.NodeBackgroundColorSet(context.Background(), &pb.NodeBackgroundColorSetRequest{NodeId: "node-1", Red: 0, Green: 255, Blue: 0, Alpha: 255})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_Groovy(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().Groovy(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.GroovyResponse{Success: true, Result: "result"}, nil)

	_, err := mockClient.Groovy(context.Background(), &pb.GroovyRequest{GroovyCode: "test"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}

func TestMockNode_Focus(t *testing.T) {
	ctrl := gomock.NewController(t)
	defer ctrl.Finish()

	mockClient := mockpb.NewMockFreeplaneClientInterface(ctrl)
	mockClient.EXPECT().FocusNode(
		gomock.Any(),
		gomock.Any(),
	).Return(&pb.FocusNodeResponse{Success: true}, nil)

	_, err := mockClient.FocusNode(context.Background(), &pb.FocusNodeRequest{NodeId: "node-1"})
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
}
